Bio/SeqIO/Interfaces.py

```diff
@@ -17,50 +17,59 @@
 
 
 class SequenceIterator:
-    """Base class for building SeqRecord iterators (DEPRECATED).
-
-    This is no longer used in the Biopython code base, instead all the parsers
-    ended up being implemented as generator functions.
+    """Base class for building SeqRecord iterators.
 
     You should write a __next__ method to return SeqRecord  objects.  You may
     wish to redefine the __init__ method as well.
     """
 
-    def __init__(self, handle, alphabet=generic_alphabet):
+    def __init__(self, source, alphabet=generic_alphabet, mode="t", fmt=None):
         """Create a SequenceIterator object.
 
         Arguments:
-        - handle - input file
+        - source - input file stream, or path to input file
         - alphabet - optional, e.g. Bio.Alphabet.generic_protein
 
-        This method MAY be overridden by any subclass, for example if you need
-        to process a header or accept additional arguments.
+        This method MAY be overridden by any subclass.
 
         Note when subclassing:
-        - there should be a single non-optional argument, the handle.
+        - there should be a single non-optional argument, the source.
         - you do not have to require an alphabet.
         - you can add additional optional arguments.
         """
-        import warnings
-        from Bio import BiopythonDeprecationWarning
-
-        warnings.warn(
-            "SequenceIterator is deprecated, and we intend to remove it"
-            " in a future release of Biopython. Please contact the Biopython"
-            " developers if you need this class.",
-            BiopythonDeprecationWarning,
-        )
-        self.handle = handle
         self.alphabet = alphabet
+        try:
+            self.stream = open(source, "r" + mode)
+            self.should_close_stream = True
+        except TypeError:  # not a path, assume we received a stream
+            if mode == "t":
+                if source.read(0) != "":
+                    raise StreamModeError(
+                        "%s files must be opened in text mode." % fmt
+                    ) from None
+            elif mode == "b":
+                if source.read(0) != b"":
+                    raise StreamModeError(
+                        "%s files must be opened in binary mode." % fmt
+                    ) from None
+            else:
+                raise ValueError("Unknown mode '%s'" % mode)
+            self.stream = source
+            self.should_close_stream = False
+        try:
+            self.records = self.parse(self.stream)
+        except Exception:
+            if self.should_close_stream:
+                self.stream.close()
+            raise
```

Comment:
Instead of having a blanket `except Exception` handler, would it not be cleaner to use `finally`? Something like:

```
try:
    return next(self.records)
finally:
    if self.should_close_stream:
        self.stream.close()
```

AFAIK it also achieves the same effect, where the unhandled exception is raised. The main difference is intent: it's clear that the `self.stream.close()` call is a cleanup call that must be executed no matter what.

Reply:
`finally` would close the file immediately after reading the first record.